File: Lista1/fa.py
import sys

# Implementacja automatu skończonego

class FiniteAutomaton:
    # Funkcja do obliczania funkcji przejścia dla danego wzorca i tekstu
    def compute_transition_function(self, pattern, text):
        m = len(pattern)  # Długość wzorca
        transitions = [{} for _ in range(m + 1)]  # Inicjalizacja tablicy przejść
        for q in range(m + 1):  # Iteracja po stanach automatu
            for a in text:  # Iteracja po znakach tekstu
                k = min(m + 1, q + 1)  # Ustawienie wartości k
                while k > 0 and not (pattern[:q] + a).endswith(pattern[:k]):  # Aktualizacja k dla funkcji przejścia
                    k -= 1
                transitions[q][a] = k  # Zapisanie wyniku do tablicy przejść
            for a in set(text) - set(transitions[q].keys()):  # Dodanie brakujących przejść
                transitions[q][a] = 0
        return transitions  # Zwrócenie tablicy przejść

    # Funkcja wyszukująca wzorzec w tekście za pomocą automatu skończonego
    def search_pattern(self, text, pattern):
        transitions = self.compute_transition_function(pattern, text)  # Obliczenie tablicy przejść
        q = 0  # Inicjalizacja stanu
        results = []  # Inicjalizacja listy wyników
        for i, c in enumerate(text):  # Iteracja po znakach tekstu wraz z indeksem
            q = transitions[q].get(c, 0)  # Aktualizacja stanu zgodnie z tablicą przejść
            print(q)
            if q == len(pattern):  # Sprawdzenie, czy osiągnięto koniec wzorca
                results.append(i - len(pattern) + 1)  # Dodanie indeksu do listy wyników
        return results  # Zwrócenie listy indeksów, na których znaleziono wzorzec

# Wywołanie programu FA
if sys.argv[1] == 'FA':  # Sprawdzenie argumentu wywołania
    fa = FiniteAutomaton()  # Inicjalizacja obiektu automatu skończonego
    with open(sys.argv[2], 'r') as pattern_file:  # Otwarcie pliku z wzorcem do odczytu
        pattern = pattern_file.readline().rstrip()  # Odczyt wzorca z pliku
    filename = sys.argv[3]  # Nazwa pliku podana jako argument wywołania
    with open(filename, 'r') as file:  # Otwarcie pliku do odczytu
        text = file.read().replace('\n', '')  # Odczyt tekstu z pliku
        results = fa.search_pattern(text, pattern)  # Wyszukanie wzorca w tekście
        print("Wzorzec " + pattern + " znaleziono na indeksach:", results)  # Wyświetlenie wyników

File: Lista1/fa.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct FiniteAutomaton
{
    char **transitions;
    int startState;
    int finalState;
};

void init(struct FiniteAutomaton *fa, char *pattern)
{
    int i, j;
    int n = strlen(pattern);
    fa->transitions = (char **)malloc((n + 1) * sizeof(char *));
    fa->startState = 0;
    fa->finalState = n;

    for (i = 0; i <= n; i++)
    {
        fa->transitions[i] = (char *)malloc(256 * sizeof(char));
        for (j = 0; j < 256; j++)
        {
            fa->transitions[i][j] = -1;
        }
    }

    for (i = 0; i < n; i++)
    {
        fa->transitions[i + 1][pattern[i]] = i + 1;
    }
}

int find(struct FiniteAutomaton *fa, char *text)
{
    int i;
    int n = strlen(text);
    int state = fa->startState;

    for (i = 0; i < n; i++)
    {
        char currentChar = text[i];
        if (currentChar < 256 && fa->transitions[state][currentChar] != -1)
        {
            state = fa->transitions[state][currentChar];
            if (state == fa->finalState)
            {
                return 1;
            }
        }
    }

    return 0;
}

int main(int argc, char *argv[])
{
    int i; // Deklaracja zmiennej 'i'
    if (argc != 3)
    {
        printf("Wywołanie: FA <wzorzec> <nazwa pliku>\n");
        exit(1);
    }

    char *pattern = argv[1];
    char *filename = argv[2];

    FILE *file = fopen(filename, "r");
    if (!file)
    {
        printf("Błąd przy otwieraniu pliku\n");
        exit(1);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);
    char *text = (char *)malloc(file_size * sizeof(char));
    fread(text, sizeof(char), file_size, file);
    fclose(file);

    struct FiniteAutomaton fa;
    init(&fa, pattern);

    if (find(&fa, text))
    {
        printf("Wzór został znaleziony w tekście.\n");
    }
    else
    {
        printf("Wzór nie został znaleziony w tekście.\n");
    }

    free(text);
    for (i = 0; i < fa.finalState; i++)
    {
        free(fa.transitions[i]);
    }
    free(fa.transitions);

    return 0;
}
